airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py

```diff
@@ -0,0 +1,463 @@
+
+    entity_prefix = "campaign"
+    enable_deleted = True
+
+    @backoff_policy
+    def _read_records(self, params: Mapping[str, Any]):
+        return self._api.account.get_campaigns(params=params)
+
+
+class AdsInsights(FBMarketingIncrementalStream):
+    """doc: https://developers.facebook.com/docs/marketing-api/insights"""
+
+    primary_key = None
+
+    ALL_ACTION_ATTRIBUTION_WINDOWS = [
+        "1d_click",
+        "7d_click",
+        "28d_click",
+        "1d_view",
+        "7d_view",
+        "28d_view",
+    ]
+
+    ALL_ACTION_BREAKDOWNS = [
+        "action_type",
+        "action_target_id",
+        "action_destination",
+    ]
+
+    MAX_WAIT_TO_START = pendulum.duration(minutes=5)
+    MAX_WAIT_TO_FINISH = pendulum.duration(minutes=30)
+    MAX_ASYNC_SLEEP = pendulum.duration(minutes=5)
+    MAX_ASYNC_JOBS = 30
+    INSIGHTS_RETENTION_PERIOD = pendulum.duration(days=37 * 30)
+
+    action_breakdowns = ALL_ACTION_BREAKDOWNS
+    level = "ad"
+    action_attribution_windows = ALL_ACTION_ATTRIBUTION_WINDOWS
+    time_increment = 1
+
+    breakdowns = []
+
+    def __init__(self, buffer_days, days_per_job, **kwargs):
+        super().__init__(**kwargs)
+        self.lookback_window = pendulum.Duration(days=buffer_days)
+        self._days_per_job = days_per_job
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        """Waits for current job to finish (slice) and yield its result"""
+        result = self.wait_for_job(stream_slice["job"])
+        # because we query `lookback_window` days before actual cursor we might get records older then cursor
+        stream_state = stream_state or {}
+        state_value = stream_state.get(self.cursor_field)
+        min_cursor = self._start_date if not state_value else pendulum.parse(state_value)
+
+        for obj in result.get_result():
+            if pendulum.parse(obj[self.cursor_field]) >= min_cursor:
+                yield obj.export_all_data()
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, Any]]]:
+        """Slice by date periods and schedule async job for each period, run at most MAX_ASYNC_JOBS jobs at the same time.
+        This solution for Async was chosen because:
+        1. we should commit state after each successful job
+        2. we should run as many job as possible before checking for result
+        3. we shouldn't proceed to consumption of the next job before previous succeed
+        """
+        stream_state = stream_state or {}
+        jobs = []
+        date_ranges = list(self._date_ranges(stream_state=stream_state))
+
+        # accumulate MAX_ASYNC_JOBS jobs in the buffer to schedule them all before trying to wait
+        for params in date_ranges[: self.MAX_ASYNC_JOBS]:
```

Comment:
shouldn't we process `MAX_ASYNC_JOBS` at a time? The current impl seems to process `MAX_ASYNC_JOBS` the first time then creates as many jobs as is left which may be greater than the max

Reply:
because we yield each job we wouldn't advance to next job before we read the result of the previous, so each `yield` is a `wait for a job`

Reply:
so this launches `MAX_ASYNC_JOBS` in parallel, waits for all of them to complete, then runs jobs one-by-one after that?